### PR TITLE
Python implementation of Spark datasource autologging

### PR DESCRIPTION
**Pull request by smurching**
## What changes are proposed in this pull request?

Depends on #2181. Adds Python `mlflow.spark.autolog()` API for autologging Spark datasource information at read-time. See the [design doc](https://docs.google.com/document/d/11nhwZtj-rps0stxuIioFBM9lkvIh_ua45cAFy_PqdHU/edit) for more details.

## How is this patch tested?
* Integration for Spark datasource autologging under `tests/spark_autologging/test_spark_datasource_autologging.py`
* Integration tests for Spark datasource autologging when used with other autologging frameworks under `tests/spark_autologging/test_spark_datasource_autologging_integrations.py`

### Is this a user-facing change?
Adds an experimental `mlflow.spark.autolog()` API for automatic logging of Spark datasource information to the current active run. Note that this API may drop datasource information if Spark is overloaded & requires installation of the mlflow-spark Java package. See the API docs for more info & usage examples.

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

### What component(s) does this PR affect?

- [ ] UI
- [ ] CLI
- [ ] API
- [ ] REST-API
- [ ] Examples
- [ ] Docs
- [x] Tracking
- [ ] Projects
- [ ] Artifacts
- [ ] Models
- [ ] Scoring
- [ ] Serving
- [ ] R
- [ ] Java
- [ ] Python

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -28,6 +28,14 @@ matrix:
           fi
     - stage: small
       if: type != cron
+    - language: python
```
Comment (smurching) [inline .travis.yml]: Will revert this change as the PR gets closer to actually merging

**Review by ColleenKeegan — APPROVED** [review]

File: mlflow/_spark_autologging.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    def tags(self):
+        with _lock:
+            global _table_infos
+            seen = set()
+            unique_infos = []
+            for info in _table_infos:
+                if info not in seen:
+                    unique_infos.append(info)
+                    seen.add(info)
+            try:
```
Comment (tomasatdatabricks) [inline mlflow/_spark_autologging.py]: Why do we need try / finally? If we expect failures, should we at least attempt to log tables that do not fail?

File: mlflow/_spark_autologging.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        from mlflow.tracking.context.registry import _run_context_provider_registry
+        _run_context_provider_registry.register(SparkAutologgingContext)
+
+
+def _get_repl_id():
+    repl_id = SparkContext.getOrCreate().getLocalProperty("spark.databricks.replId")
+    if repl_id:
+        return repl_id
+    main_file = sys.argv[0] if len(sys.argv) > 0 else "<console>"
+    return "PythonSubscriber[{filename}][{id}]".format(filename=main_file, id=uuid.uuid4().hex)
```
Comment (tomasatdatabricks) [inline mlflow/_spark_autologging.py]: Can you document what is the repl id for and when do you generate a new one?

File: tests/spark_autologging/utils.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        Row(-27, 55, "horse")
+    ]
+    schema = StructType([
+        StructField("number2", IntegerType()),
+        StructField("number1", IntegerType()),
+        StructField("word", StringType())
+    ])
+    rdd = spark_session.sparkContext.parallelize(rows)
+    df = spark_session.createDataFrame(rdd, schema)
+    format_to_file_path = {}
```
Comment (tomasatdatabricks) [inline tests/spark_autologging/utils.py]: nit: name this something else to avoid confusion with the function name?

**Review by tomasatdatabricks — COMMENTED** [review]
Looks good! I had just a few minor comments.

File: mlflow/_spark_autologging.py (hunk 2 above)
Comment (smurching) [reply mlflow/_spark_autologging.py]: Good question, we don't need the try-finally, I'll simplify this

**Review by tomasatdatabricks — APPROVED** [review]
lgtm!
